File: tests/ops/test_div.py
import torch
import torch_webgpu  # noqa: F401


def test_div_basic():
    a = torch.tensor([10.0, 20.0, 30.0], device="webgpu")
    b = torch.tensor([2.0, 4.0, 5.0], device="webgpu")
    result = a / b
    expected = torch.tensor([5.0, 5.0, 6.0])
    assert torch.allclose(result.to("cpu"), expected)


def test_div_fractional():
    a = torch.tensor([1.0, 3.0, 5.0], device="webgpu")
    b = torch.tensor([3.0, 7.0, 2.0], device="webgpu")
    result = a / b
    expected = a.to("cpu") / b.to("cpu")
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-4, atol=1e-4)


def test_div_negative():
    a = torch.tensor([-10.0, 20.0, -30.0], device="webgpu")
    b = torch.tensor([2.0, -4.0, -5.0], device="webgpu")
    result = a / b
    expected = torch.tensor([-5.0, -5.0, 6.0])
    assert torch.allclose(result.to("cpu"), expected)


def test_div_broadcast():
    # Create on CPU first, then move to WebGPU (randn not implemented on WebGPU)
    a_cpu = torch.randn(3, 4) + 1.0
    b_cpu = torch.randn(4).abs() + 0.1
    a = a_cpu.to("webgpu")
    b = b_cpu.to("webgpu")
    result = a / b
    expected = a_cpu / b_cpu
    assert torch.allclose(result.to("cpu"), expected, rtol=1e-4, atol=1e-4)

File: ctests/test_div.cpp
#include <gtest/gtest.h>
#include <torch/torch.h>

namespace
{

    torch::Device webgpu_device()
    {
        return torch::Device(torch::DeviceType::PrivateUse1);
    }

    torch::Tensor to_webgpu(const torch::Tensor &cpu)
    {
        return cpu.to(webgpu_device());
    }

} // namespace

TEST(ArithmeticOps, DivBasic)
{
    auto a_cpu = torch::tensor({10.0f, 20.0f, 30.0f});
    auto b_cpu = torch::tensor({2.0f, 4.0f, 5.0f});
    auto a = to_webgpu(a_cpu);
    auto b = to_webgpu(b_cpu);

    auto out = torch::div(a, b);
    auto expected = torch::div(a_cpu, b_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected));
}

TEST(ArithmeticOps, DivFractional)
{
    auto a_cpu = torch::tensor({1.0f, 3.0f, 5.0f});
    auto b_cpu = torch::tensor({3.0f, 7.0f, 2.0f});
    auto a = to_webgpu(a_cpu);
    auto b = to_webgpu(b_cpu);

    auto out = torch::div(a, b);
    auto expected = torch::div(a_cpu, b_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-4, 1e-4));
}

TEST(ArithmeticOps, DivNegative)
{
    auto a_cpu = torch::tensor({-10.0f, 20.0f, -30.0f});
    auto b_cpu = torch::tensor({2.0f, -4.0f, -5.0f});
    auto a = to_webgpu(a_cpu);
    auto b = to_webgpu(b_cpu);

    auto out = torch::div(a, b);
    auto expected = torch::div(a_cpu, b_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected));
}

TEST(ArithmeticOps, DivBroadcast)
{
    auto a_cpu = torch::randn({3, 4}) + 1.0f;
    auto b_cpu = torch::randn({4}).abs() + 0.1f;
    auto a = to_webgpu(a_cpu);
    auto b = to_webgpu(b_cpu);

    auto out = torch::div(a, b);
    auto expected = torch::div(a_cpu, b_cpu);
    ASSERT_TRUE(torch::allclose(out.to(torch::kCPU), expected, 1e-4, 1e-4));
}
